### DISCUSSION TITLE
docs on what steps the controller follows

### DISCUSSION THREAD

**Discussion by tooptoop4** [argoproj/argo-workflows#13729] [Q&A] [answered]
is it documented what the order of operations/steps of the controller are doing? 
for a (retry/steps not dag workflow). ie 

1. how does it detect a pod has finished doing work
2. what it does to the pod next (send terminate/kill/delete?)
3. what it needs to see before it can mark the node success
4. what it needs to see before it can mark the workflow success
5. how workflowtaskresults fit in?
6. would really like to know how it knows a node succeded at 12.03pm but it only updates that node's endtime at 12.06pm (confirmed by dumping the workflow yaml every 20s and seeing no mention of that 12.03pm node endtime until 12.06pm)
7. instead of sending terminate then kill then delete to a pod, why can't it just do delete first

**Comment by agilgur5** [comment, marked as answer]
I saw you asked this [on Slack](https://cloud-native.slack.com/archives/C0510EUH90V/p1728161624856229) too. I wrote about this partly in https://github.com/argoproj/argo-workflows/issues/12800#issuecomment-1998956903, the "Executor" section. This is not well documented but is also pretty low-level details as I wrote there, as it's meant to be as transparent as possible to an ordinary user.

1. Emissary is a parent process of your container's `command`. That's why the [entrypoint lookup feature](https://argo-workflows.readthedocs.io/en/release-3.5/workflow-executors/#image-indexcache) exists. There's a pretty good description of how the `init`, `main`, and `wait` interact [in one of the Executor files](https://github.com/argoproj/argo-workflows/blob/ceaabf12cb86f185eb89c3e69880f20d61dfe6af/workflow/executor/emissary/emissary.go#L23-L45). Emissary detects the sub process exiting as completion and it communicates that completion to the Controller via `WorkflowTaskResults` (previously by `pod` `patch`ing annotations: #13100)). This communication may fail due to a non-graceful exit (those are most of the 3.5.3-3.5.10 issues like #12993 et al).
1. Emissary will then exit gracefully on its own. Signals are only sent if imperative "Stop" or "Terminate" are done or `activeDeadlineSeconds` is hit -- i.e. only if you've told the Worflow to stop at some point.
1. The Controller needs to see a completed WorkflowTaskResult
1. Based on node successes. There do seem to be some unexpected results or bugs in the algorithm here though, like https://github.com/argoproj/argo-workflows/issues/12530 et al 
1. See above
1. I'm not sure, it depends on Informer/k8s API latency as well on graceful vs non-graceful exit (https://github.com/argoproj/argo-workflows/pull/13454 has a 2 min non-graceful timeout). Some older versions had Informer `resourceVersion` issues as well (https://github.com/argoproj/argo-workflows/pull/12133 et al). Would really need more details on that.
1. Pods are their own k8s resource and are needed for live logs amongst other things. `podGC`, workflow TTL etc, handles deletion of Pods instead based on your configurations. See also https://github.com/argoproj/argo-workflows/issues/13307#issuecomment-2215415790

**Reply by tooptoop4** [reply]
thx, i always assumed based on log timestamps that the workflow needed to delete the pod of an earlier node to be able to progress to next node but based on above seems workflowtaskresult might be key to my ongoing issue.

 your info led me to discover:
https://github.com/argoproj/argo-workflows/pull/8000
https://github.com/argoproj/argo-workflows/pull/8135
https://github.com/argoproj/argo-workflows/issues/10096
https://github.com/argoproj/argo-workflows/pull/10097
https://github.com/argoproj/argo-workflows/pull/13491
https://github.com/argoproj/argo-workflows/pull/13454
https://github.com/argoproj/argo-workflows/issues/13344
https://github.com/argoproj/argo-workflows/pull/13537

and find another informer in https://github.com/argoproj/argo-workflows/issues/13671#issuecomment-2403350078 , hopefully this is the last piece of the puzzle
